packages/kbn-gateway/src/config/config_service.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { get } from 'lodash';
+import { firstValueFrom } from 'rxjs';
+import { fromRoot } from '@kbn/utils';
+import { RawConfigService, ConfigPath } from '@kbn/config';
+import { Type } from '@kbn/config-schema';
+import { getArgValues } from './read_argv';
+
+const CONFIG_CLI_FLAGS = ['-c', '--config'];
+const DEFAULT_CONFIG_PATH = fromRoot('config/gateway.yml');
+
+const pathToString = (path: ConfigPath) => (Array.isArray(path) ? path.join('.') : path);
+
+export interface ConfigStart {
+  /** Signature borrowed from the method of the same name in core's own ConfigService. */
+  atPathSync: <TSchema>(path: ConfigPath) => TSchema;
+}
+
+/**
+ * A very simple service for loading configuration from a file, validating it,
+ * and providing a means for other services to retrieve it.
+ */
+export class ConfigService {
+  private readonly rawConfigService: RawConfigService;
+  private readonly schemas = new Map<string, Type<unknown>>();
+  private readonly validatedConfigs = new Map<string, unknown>();
+
+  constructor() {
+    const configPathOverride = getArgValues(process.argv, CONFIG_CLI_FLAGS);
+    const configPath = configPathOverride.length ? configPathOverride : [DEFAULT_CONFIG_PATH];
+    this.rawConfigService = new RawConfigService(configPath);
+  }
+
+  public setSchema(path: ConfigPath, schema: Type<unknown>) {
+    const namespace = pathToString(path);
+    if (this.schemas.has(namespace)) {
+      throw new Error(`Validation schema for [${path}] was already registered`);
+    }
+
+    this.schemas.set(namespace, schema);
+  }
+
+  async start(): Promise<ConfigStart> {
+    const config = (await this.loadConfig()) as Record<string, unknown>;
+
+    // Validate and cache all configs.
+    for (const namespace of this.schemas.keys()) {
+      this.validateAtPath(namespace, config);
+    }
```

Comment:
Since we don't anticipate a need to support config reloads, we can KISS and just depend on everything being validated once up-front.